pkg/controller/common/certificates/secret.go

```diff
@@ -63,29 +66,78 @@ func HTTPCertSecretVolume(namer name.Namer, name string) volume.SecretVolume {
 	)
 }
 
-type CertificatesSecret v1.Secret
+type CertificatesSecret struct {
+	v1.Secret
+	ca *CA
+}
+
+func NewCertificatesSecret(secret v1.Secret) (*CertificatesSecret, error) {
+	result := CertificatesSecret{Secret: secret}
+	if err := result.validate(); err != nil {
+		return nil, err
+	}
+	return &result, nil
+}
 
 // CAPem returns the certificate of the certificate authority.
-func (s CertificatesSecret) CAPem() []byte {
+func (s *CertificatesSecret) CAPem() []byte {
 	return s.Data[CAFileName]
 }
 
+// CA returns a pointer to the in-memory representation of the CA contained in the secret.
+func (s *CertificatesSecret) CA() *CA {
+	return s.ca
+}
+
 // CertChain combines the certificate of the CA and the host certificate.
-func (s CertificatesSecret) CertChain() []byte {
+func (s *CertificatesSecret) CertChain() []byte {
 	return append(s.CertPem(), s.CAPem()...)
 }
 
-func (s CertificatesSecret) CertPem() []byte {
+func (s *CertificatesSecret) CertPem() []byte {
 	return s.Data[CertFileName]
 }
 
-func (s CertificatesSecret) KeyPem() []byte {
+func (s *CertificatesSecret) KeyPem() []byte {
 	return s.Data[KeyFileName]
 }
 
-// Validate checks that mandatory fields are present.
+func (s *CertificatesSecret) HasFullCA() bool {
+	if s == nil {
+		return false
+	}
+	// the presence of the key means by implication that we have a full CA, validation ensures that the CA cert exists
+	_, exists := s.Data[CAKeyFileName]
+	return exists
+}
+
+func (s *CertificatesSecret) HasLeafCertificate() bool {
+	return s != nil && !s.HasFullCA()
+}
+
+func (s *CertificatesSecret) validateCustomCA() error {
+	// flag up user error when specifying both CA certificate with key and leaf certificate
+	_, tlsKeyExists := s.Data[KeyFileName]
+	_, tlsCertExists := s.Data[CertFileName]
+	if tlsKeyExists || tlsCertExists {
+		return fmt.Errorf("cannot specify both %s and %s or %s in %s/%s",
```

Comment:
Took me a second to understand what `and` and `or` operate on :) Maybe something like below (+ change arguments order)?

```suggestion
		return fmt.Errorf("cannot specify %s or %s when %s is set in %s/%s",
```